src/xarray_regrid/methods/conservative.py

```diff
@@ -200,34 +226,46 @@ def apply_weights(
     )
 
 
-def compute_einsum_dask(da: xr.DataArray, weights: np.ndarray) -> dask.array.Array:
-    """Compute the einsum between dask data and weights, and mask NaNs if needed."""
-    new_data: dask.array.Array
-    if np.any(np.isnan(da.data)):
-        new_data = dask.array.einsum(
-            "i...,ij->j...", da.fillna(0).data, weights, optimize="greedy"
-        )
-        isnan = dask.array.einsum(
-            "i...,ij->j...", np.isnan(da.data), weights, optimize="greedy"
-        )
-        new_data[isnan > 0] = np.nan
+def compute_einsum(
+    da: xr.DataArray, weights: np.ndarray, skipna: bool, nan_threshold: float
+) -> np.ndarray | dask.array.Array:
+    """Compute the einsum between the data and weights. If nan-skipping is
+    enabled, renormalize the weights by the non-null input points, and keep
+    output points containing up to the specified fraction of NaN values."""
+    new_data: np.ndarray | dask.array.Array
+
+    if skipna:
+        data = da.fillna(0).data
+        notnull = da.notnull().data
     else:
-        new_data = dask.array.einsum(
-            "i...,ij->j...", da.data, weights, optimize="greedy"
+        data = da.data
+
+    backend = np if da.chunks is None else dask.array
+    new_data = backend.einsum(
+        "i...,ij->j...", data, weights, optimize="greedy"
+    )
+
+    if skipna:
+        valid_weight_sum = backend.einsum(
+            "i...,ij->j...", notnull, weights, optimize="greedy"
         )
+        new_data = new_data / backend.clip(valid_weight_sum, 1e-6, None)
+        weight_threshold = get_weight_threshold(nan_threshold)
+        new_data = backend.where(valid_weight_sum > weight_threshold, new_data, np.nan)
```

Comment:
Thanks for implementing this! I wasn't sure at first how viable this would be. Did you look at the memory overhead here for processing Dask data?